### PR TITLE
Handle slashes in oracle schema/table names #13225

### PR DESCRIPTION
**Pull request by dpsutton**
table names like `table_slash/` are valid in our normal query
apparatus but are invalid in the calls to `.getColumns` and
`.getTables` when determining the shape of the database. In these
contexts it needs to be `table_slash//`

- also get fix TODO to reduce over the list of tables rather than
realizing entire result set and then reducing

###### Before submitting the PR, please make sure you do the following
- [ ] If you're attempting to fix a translation issue, please submit your changes to our [POEditor project](https://poeditor.com/join/project/ynjQmwSsGh) instead of opening a PR.
### Tests
-  [ ] Run the frontend and integration tests with  `yarn lint && yarn flow && yarn test`)
-  [x] If there are changes to the backend codebase, run the backend tests with `lein test && lein lint && ./bin/reflection-linter`

-  [x] Sign the [Contributor License Agreement](https://docs.google.com/a/metabase.com/forms/d/1oV38o7b9ONFSwuzwmERRMi9SYrhYeOrkbmNaq9pOJ_E/viewform)
(unless it's a tiny documentation change).

### CODE REVIEW

**Comment by dpsutton** [comment]
Debated a bit about where the `metadata-escape-entity-name` should go. It's really only specific to jdbc based drivers and i thought about putting it in `metabase.driver.sql-jdbc.common`. However, it wouldn't have access to the same hierarchy and dispatch already used in the other main `driver` namespace (albeit the driver would assuredly have already been loaded at this point) but it still felt like it might be a bit surprising.

I'm also still investigating how things work when upgrading. IE, we've indexed these tables and they aren't working and then the patch here is installed. It seems like our mechanism for updating our metadata might not be as strong as we'd like and it might be preferable to remove the database and re-add it. Still investigating.

**Comment by sbelak** [comment]
If I understand what's happening in #13225 correctly, sync will have failed and no tables from problematic schemas are added. 

As far as I'm concerned, code organization is fine. 

P.S. to trigger driver-specific CI runs, append `[ci drivers]` to the commit message

**Comment by dpsutton** [comment]
Adding tests don't merge.

**Comment by dpsutton** [comment]
@sbelak Can you review this for me please? The mongo and bigquery failures are unreleated, but might be related to the `:type/Structured` change. I'm investigating now

File: modules/drivers/oracle/src/metabase/driver/oracle.clj
Comment (sbelak) [inline modules/drivers/oracle/src/metabase/driver/oracle.clj:305]: We usually just use `_`. This is super minor, but when I first read for a moment I though that the fn takes 3 args.

Also perhaps `table-or-schema` or `entity-name` would be names that better match how the fn is used.

File: modules/drivers/oracle/test/metabase/driver/oracle_test.clj
Comment (sbelak) [inline modules/drivers/oracle/test/metabase/driver/oracle_test.clj:165]: Can you add a test for schema as well

File: src/metabase/driver/sql_jdbc/sync.clj
Comment (sbelak) [inline src/metabase/driver/sql_jdbc/sync.clj:103]: This could be even more concisely written using `into` (while retaining transducer semantics):
```
(into [] 
  (map (fn [row] (select-keys row [:table_name :remarks :table_schem]))) 
  (jdbc/reducible-result-set rs {}))
```

File: modules/drivers/oracle/src/metabase/driver/oracle.clj
Comment (sbelak) [inline modules/drivers/oracle/src/metabase/driver/oracle.clj:304]: We usually go for the more narrative names like `escape-entity-name-for-metadata` or `-for-sync` rather than prefixing (took me a while to get comfortable with this style -- and sometimes I still lapse -- but am always thankful for it when reading code I'm not intimate with).

**Review by sbelak — APPROVED** [review]
Looks good. Minor stylistic nitpicks

**Comment by sbelak** [comment]
The failing tests certainly look like they were caused by the fingerprinting/OOM fix. I see that master is broken too for a while for various reasons which is why we didn't picked up on it sooner.

**Comment by dpsutton** [comment]
ah. it looks like bigquery and mongo can't handle expressions (which we are doing to get substrings)

:cause "No method in multimethod '->lvalue' for dispatch value: :expression"
 :via
 [{:type java.lang.IllegalArgumentException
   :message "No method in multimethod '->lvalue' for dispatch value: :expression"
   :at [clojure.lang.MultiFn getFn "MultiFn.java" 156]}]

**Comment by sbelak** [comment]
True, although I don't see why BigQuery doesn't. Either way, you can check for supported features using 
`(driver/supports? driver/*driver* :expressions)`

**Comment by flamber** [comment]
I went over all the `driver/supports?` in #12675 and also didn't know why BigQuery didn't have `:expressions`

**Comment by dpsutton** [comment]
@sbelak can you tell me how to tag this PR properly before merging? I want to make sure it gets the right metadata added to it

**Pull request merged**
